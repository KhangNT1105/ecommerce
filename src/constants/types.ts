export type MarketType = {
  countryCode: string;
  countryName: string;
  currencyCode: string;
  currencySymbol: string;
  active: boolean;
};

export type ClientType = {
  clientId: string;
  clientName: string;
  active: boolean;
};

export type PaginationType = {
  totalPages: number;
  totalItems: number;
  hasPrevPage: boolean;
  hasNextPage: boolean;
  prevPage: number | null;
  nextPage: number | undefined;
  size: number | 10;
  page: number | undefined;
};

export type MenuItem = {
  path: string;
  name: string;
  icon: string;
  children: MenuItem[];
};

export type ProductDenominationType = {
  label: string;
  price: number;
};

export interface IProductImage {
  logo: string;
  thumbnail?: string;
  portrait?: string;
  landscape?: string;
}

export interface ICategory {
  id: string;
  name: string;
  parentCategoryId?: string;
  externalCategoryId?: string;
  vendor?: string;
}

export interface IProductItem {
  id: string;
  displayName: string;
  vendor: string;
  productId?: string;
  market?: string;
  productBrand?: string;
  ghostLinkExpiryDays?: number;
  minimumAllowableExpiryDays?: number;
  valueType?: string;
  minPrice?: number;
  maxPrice?: number;
  redemptionType?: string;
  denominations?: string[];
  productDenominations?: ProductDenominationType[];
  ghostLinkInMarket?: number;
  allocated?: number;
  inQueue?: number;
  stockOnHand?: number;
  expiring?: number;
  internalName?: string;
  originalName?: string;
  description?: string;
  externalProductId?: string;
  categoryIds?: string[];
  published?: string;
  outOfStockBehaviour?: string;
  images?: IProductImage;
  brand?: {
    brandLogo?: string;
    brandName?: string;
  };
}

export interface AnyObject {
  [key: string]: any;
}

export interface StringPropertyObject {
  [key: string]: string;
}

export const SortValues = {
  ASC: 'updatedAt.asc',
  DESC: 'updatedAt.desc'
};

export const PortalType = {
  DOMAIN: 'domain',
  SMS: 'sms'
};

export type FnType = (param?: object) => void;
export type FnSubmitFormType = (event: React.FormEvent<HTMLFormElement>) => void;

export type EXPORT_TYPE = 'txt' | 'json' | 'csv' | 'xls' | 'xml';

export type PathParamsWithIDType = {
  id: string;
};
