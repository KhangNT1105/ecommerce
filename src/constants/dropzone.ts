import Configs from 'config';

export const UPLOAD_ACCEPT_FILE_TYPE: string = 'image/jpeg, image/jpg, image/png, image/svg, image/bmp, image/gif';
export const UPLOAD_ACCEPT_FILE_VIDEO_TYPE: string =
  'video/mp4, video/webm, video/wmv, video/flv, video/ogv, video/avi, video/mov';
export const FILE_DROPZONE_MIN_SIZE: number = 0;
export const FILE_DROPZONE_MAX_SIZE: number = 10048576;
export const FILE_DROPZONE_VIDEO_MAX_SIZE: number = 1000048576;
export const FILE_SIZE_IN_MB: number = 1;
export const FILE_SIZE_IN_PX: number = 1920;
export const RESOLUTION_MIN_SIZE_LANDSCAPE: number = 0;
export const RESOLUTION_MIN_SIZE_PORTRAIT: number = 0;

export const defaultDropzone = {
  apiURL: Configs.API.UPLOAD_SERVICE,
  files: [],
  autoUpload: true,
  config: {
    accept: UPLOAD_ACCEPT_FILE_TYPE,
    disabled: false,
    noClick: false,
    noKeyboard: false,
    minSize: FILE_DROPZONE_MIN_SIZE,
    maxSize: FILE_DROPZONE_MAX_SIZE
  },
  pushData: {
    error: false,
    msg: 'OK',
    infos: {
      title: 'Hero Banner Desktop and big screens',
      url: 'https://i.pinimg.com/236x/0a/8b/fb/0a8bfb162ef8a917f1e4d4bf5cd08e4c--android-apps-cover-art.jpg',
      size: '12222'
    }
  }
};

export const defaultDropzoneVideo = {
  ...defaultDropzone,
  config: {
    ...defaultDropzone.config,
    accept: UPLOAD_ACCEPT_FILE_VIDEO_TYPE,
    maxSize: FILE_DROPZONE_VIDEO_MAX_SIZE
  }
};

export const defaultDropzonePreviewInit = {
  files: [],
  onFileRemove: () => {},
  onFileUpload: () => {},
  status: false
};
