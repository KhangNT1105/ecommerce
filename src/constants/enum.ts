// TODO: All enum should be listed here
import { TypeOptions, ToastPosition } from 'react-toastify';
import { EXPORT_TYPE } from './types';

export const COLOR = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning',
  DANGER: 'danger',
  LINK: 'link'
};

export const SIZE = {
  LARGE: 'lg',
  SMALL: 'sm'
};

export const DURATION = {
  TOAST: 5000,
  TOAST_TRANSITION: 1000
};

interface POSITION_OPTIONS {
  [pos: string]: ToastPosition;
}
export const POSITION: POSITION_OPTIONS = {
  TOP_LEFT: 'top-left',
  TOP_RIGHT: 'top-right',
  TOP_CENTER: 'top-center',
  BOTTOM_LEFT: 'bottom-left',
  BOTTOM_RIGHT: 'bottom-right',
  BOTTOM_CENTER: 'bottom-center'
};

interface TYPE_OPTIONS {
  [type: string]: TypeOptions;
}
export const TOAST_TYPE: TYPE_OPTIONS = {
  SUCCESS: 'success',
  ERROR: 'error'
};

export const DATE_RANGE = {
  TODAY: 'today',
  LAST_7_DAYS: 'last7days',
  LAST_30_DAYS: 'last30days',
  LAST_WEEK: 'lastweek',
  LAST_MONTH: 'lastmonth',
  CUSTOM_DATE: 'customdate'
};

interface FILE_TYPE {
  [pos: string]: EXPORT_TYPE;
}
export const EXPORT_FILE_TYPE: FILE_TYPE = {
  CSV: 'csv',
  XLS: 'xls'
};

export const BREAKPOINTS = {
  EXTRA_LARGE: 1200,
  LARGE: 992,
  MEDIUM: 768,
  SMALL: 576
};

export const DROPZONE_EVENT_TYPE = {
  DROP: 'drop',
  CLICK: 'click'
};

export const ENTITIES = {
  PATH: {
    CONTENT: 'content',
    LICENSES: 'licenses',
    USER_MANAGEMENT: 'user-management',
    SYSTEM_CONFIGURATION: 'system-configuration'
  },
  NAME: {
    SHOWS: 'shows',
    SCHEDULES: 'schedules',
    CHANNELS: 'channels',
    CATALOGS: 'catalogs',
    CATEGORIES: 'categories',
    CASTINGS: 'casts',
    CLIPS: 'clips',
    PLAYLISTS: 'playlists',
    UPDATES: 'updates',
    CONTRACTS: 'contracts',
    LICENSORS: 'licensors',
    SYSTEM_USERS: 'system-users',
    USER_GROUPS: 'user-groups',
    SEASONS: 'seasons',
    CONFIGURATIONS: 'configuration',
    EPISODES: 'episodes',
    PAYMENT_GATEWAYS: 'payment-gateways',
    RISK_ENGINE_RULES_CONFIG: 'risk-engine'
  }
};

export const PUBLIC_PAGES = {
  NAME: {
    WELCOME: '/',
    LOGIN: 'login',
    TAC: 'term-and-conditions',
    ACCESS_DENIED: 'access-denied-403',
    DEMO: '/demo',
    PRODUCTS: 'sanpham',
    PRODUCT: 'sanpham/:id',
    CART: 'cart',
    CHECKOUT: 'checkout'
  }
};

export const LANGUAGES_CODE = {
  VN: 'vn',
  EN: 'en'
};

// This is API's entities from BE
export const API_ENTITIES_NAME = {
  SHOWS: 'shows',
  SCHEDULES: 'schedules',
  CHANNELS: 'channels',
  CATALOGS: 'catalogs',
  CATEGORIES: 'categories',
  CASTINGS: 'casts',
  CLIPS: 'clips',
  PLAYLISTS: 'playlists',
  UPDATES: 'updates',
  CONTRACTS: 'contracts',
  LICENSORS: 'licensors',
  SYSTEM_USERS: 'usergroups', // BE don't have entity's name : users | using same usergroups entity
  USER_GROUPS: 'usergroups',
  SEASONS: 'seasons',
  CONFIGURATIONS: 'configurations',
  EPISODES: 'episodes',
  RISK_ENGINES: 'riskengines'
};

export const PAGE_ACTIONS = {
  CREATE: 'create',
  EDIT: 'edit',
  DELETE: 'delete'
};

export const API_ERROR_CODE = {
  TOKEN_INVALID: '1012',
  TOKEN_EXPIRED: '1011',
  REFRESH_TOKEN_EXPIRED: '1014',
  REFRESH_TOKEN_INVALID: '1015'
};

export const PERMISSION_PREFIX = {
  VIEW: 'VIEW',
  CREATE: 'CREATE',
  EDIT: 'EDIT',
  EDIT_APPROVAL: 'EDIT_APPROVAL',
  EDIT_VIEW_CONTRACT: 'EDIT_VIEW_CONTRACT',
  ACTIVATE: 'ACTIVATE',
  DEACTIVATE: 'DEACTIVATE',
  APPROVE: 'APPROVE',
  REJECT: 'REJECT',
  DELETE: 'DELETE',
  RETRIEVE_BCM: 'RETRIEVE_BCM'
};

export const PAGE_ENTITY = {
  HOME: 'HOME'
};

export const BUTTONS_VISIBILITY = {
  SHOW: 'show'
};
export const PAGE_ACCESSIBILITY = {
  SHOW: 'show'
};

export const CLASS_NAME = {
  READONLY: 'read-only',
  DISABLED: 'disabled'
};

export const SEARCH_BOX_SUBMIT_DELAY = 600;

export const SHOWS_TYPES = {
  MOVIE: 'movie',
  SERIES: 'series',
  LIFE_EVENT: 'life event',
  THEATRICAL_PLAY: 'theatrical play',
  PROGRAM: 'program',
  MATCH: 'match',
  SEASONS: 'seasons',
  CHANNELS: 'channels',
  SCHEDULED: 'scheduled',
  SCHEDULING: 'ready for scheduling'
};

export const IMAGE_TYPES = {
  POSTER: 'poster',
  LANDSCAPE: 'landscape',
  POSTER_CLEAN: 'poster clean',
  LANDSCAPE_CLEAN: 'landscape clean',
  POSTER_WITH_LOGO: 'poster with logo',
  LANDSCAPE_WITH_LOGO: 'landscape with logo',
  LOGO_TITLE: 'logo title',
  HERO_BANNER_DESKTOP: 'hero banner desktop',
  HERO_BANNER_MOBILE: 'hero banner mobile'
};

export const INPUT_FIELD = {
  MAX_LENGTH_NORMAL: 100,
  MAX_LENGTH_URL: 1000
};

export const SORTING_DIRECTION = {
  ASCENDING: 'asc',
  DESCENDING: 'desc'
};
const SECONDS = {
  MINUTE: 60,
  HOUR: 3600,
  DAY: 86400,
  MONTH: 2592000,
  YEAR: 31536000
};
const ENUMS = {
  COLOR,
  SIZE,
  DURATION,
  POSITION,
  TOAST_TYPE,
  DATE_RANGE,
  EXPORT_FILE_TYPE,
  BREAKPOINTS,
  DROPZONE_EVENT_TYPE,
  SECONDS
};

export default ENUMS;
