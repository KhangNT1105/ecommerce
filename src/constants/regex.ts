const REGEX = {
  NUMBER_ONLY: /^[0-9]([.,][0-9])?$/,
  DIGIT_ONLY: /^[0-9]+$/
};

export default REGEX;
