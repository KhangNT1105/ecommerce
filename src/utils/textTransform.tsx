export const upperCaseFirst = (str: String = '') => {
  return str.toLowerCase().replace(/^\w/, (chars) => chars.toUpperCase());
};
