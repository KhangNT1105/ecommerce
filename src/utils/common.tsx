import React from 'react';
import { Denomination } from 'Props.d';
import { assign, clone, map, reduce, isObjectLike } from 'lodash';
import { validateRequired, validateEmail } from 'utils/validation';
import { ACTIVATE, ENTITY_STATUS_FROM_SERVER, DEACTIVATE, DELETE, APPROVE, REJECT } from '../constants';
import { Genre, SubGenre } from 'stores/ConfigurationStore/configurationType.d';

export const checkRenderDoubleValue = (value1?: string | number, value2?: string | number, value3?: any | '') => {
  let minus = '';
  if (value1 && value2) {
    minus = ' - ';
  }
  return (
    <>
      {value1} {value3 && <b>{value3}</b>}
      {minus}
      {value2}
    </>
  );
};

export type SelectOptionProps = {
  id: string;
  value: string;
  label: string;
};

type DataProps = {
  [key: string]: any;
};

export type SelectOptionKeyMappingProps = {
  key: string;
  value: string;
  label: string;
};

export const renderSelectOptions = (data: DataProps[], keyMapping?: SelectOptionKeyMappingProps) => {
  return (
    <>
      <option value="">Select</option>
      {data.map((item) => {
        if (keyMapping && Object.entries(keyMapping).length > 0) {
          return (
            <option key={item[keyMapping.key]} value={item[keyMapping.value]}>
              {item[keyMapping.label]}
            </option>
          );
        }
        return (
          <option key={item.id} value={item.value}>
            {item.label}
          </option>
        );
      })}
    </>
  );
};

export const checkRenderArray = (items: string[]) => {
  return <span>{items.join(', ')}</span>;
};

export const getListPrices = (denominations: Denomination[]) => {
  const listPrices = denominations.map((item) => item.price);
  return listPrices.sort((a, b) => a - b);
};
export const getPrices = (productDetail: any) => {
  if (productDetail.valueType === 'RANGE') {
    return checkRenderDoubleValue(productDetail.minPrice, productDetail.maxPrice);
  }
  if (productDetail.valueType === 'SLAB') {
    const values = getListPrices(productDetail.denominations);
    return values.join(', ');
  }
};

export interface IPagingModelOption {
  pageIndex: number;
  pageSize: number;
}

export interface IPagingModelResult<T> {
  items: T[];
  hasPrevPage: boolean;
  hasNextPage: boolean;
  prevPage: number | null;
  nextPage: number | null;
  size: number;
  page: number;
  totalItems: number;
  totalPages: number;
}

export function getPaginationResponse<T>(
  pagingModel: IPagingModelOption,
  result: T[],
  totalItems: number
): IPagingModelResult<T> {
  const totalPages = Math.floor((totalItems + pagingModel.pageSize - 1) / pagingModel.pageSize);
  const pagingModelResult: IPagingModelResult<T> = {
    items: result,
    totalItems,
    totalPages,
    hasPrevPage: pagingModel.pageIndex > 0,
    hasNextPage: pagingModel.pageIndex < totalPages - 1,
    prevPage: pagingModel.pageIndex > 0 ? pagingModel.pageIndex - 1 : null,
    nextPage: pagingModel.pageIndex < totalPages - 1 ? pagingModel.pageIndex + 1 : null,
    size: pagingModel.pageSize,
    page: pagingModel.pageIndex
  };
  return pagingModelResult;
}

export function replaceSpecialCharacters(str: string, character: string) {
  return str.replace(/(?:\.(?![^.]+$)|[^\w.])+/g, character);
}

export const classConcat = (classes: string[]): string => {
  if (classes.length) {
    return classes.filter((className) => !!className).join(' ');
  }
  return '';
};

export const recursiveGetFieldName = (obj: object, entityField: string, fieldNameCombine?: string): any => {
  let listFieldConfig: any;
  listFieldConfig = reduce(
    obj,
    (result, value, key) => {
      let cloneResult = clone(result);
      let localFieldName: any = clone(fieldNameCombine);
      if (key === entityField) {
        map(value, (item: any, k) => {
          const getFieldName = `${localFieldName}.${key}.${k}`;

          // check whether any validation
          if (item.validationRules && item.validationRules.length > 0) {
            item.validationRules.map((itemValidate: any, index: number) => {
              if (itemValidate.rule === 'isRequired') {
                item.validationRules[index].validator = validateRequired;
              }
              if (itemValidate.rule === 'isEmail') {
                item.validationRules[index].validator = validateEmail;
              }

              return itemValidate;
            });
          }

          // insert key, value into list field
          return (cloneResult = {
            ...cloneResult,
            [getFieldName]: item
          });
        }); // reset field name

        localFieldName = '';
      } else if (isObjectLike(value)) {
        if (!localFieldName) {
          localFieldName = key;
        } else {
          localFieldName += `.${key}`;
        } // recall recursive fn if not be found the expected entityField

        return assign(cloneResult, recursiveGetFieldName(value, entityField, localFieldName));
      }

      listFieldConfig = {
        ...listFieldConfig,
        ...cloneResult
      };

      return listFieldConfig;
    },
    {}
  );

  return listFieldConfig;
};

export type ObjectFieldType = {
  [field: string]: string | object | string[] | object[] | any;
};

export const trimAllValueWithObjectField = (objField: ObjectFieldType) => {
  if ((!Array.isArray(objField) && typeof objField !== 'object') || objField === null) {
    return objField;
  }
  return Object.keys(objField).reduce(
    (acc: ObjectFieldType, key) => {
      acc[key] =
        typeof objField[key] === 'string'
          ? (objField[key] as string).trim()
          : trimAllValueWithObjectField(objField[key]);
      return acc;
    },
    Array.isArray(objField) ? [] : {}
  );
};

const listShowBtnActivate = [
  ENTITY_STATUS_FROM_SERVER.INACTIVATE,
  ENTITY_STATUS_FROM_SERVER.DRAFT,
  ENTITY_STATUS_FROM_SERVER.UPDATED,
  ENTITY_STATUS_FROM_SERVER.ACTIVATE
];

const listShowBtnSMActivate = [
  ENTITY_STATUS_FROM_SERVER.INACTIVATE,
  ENTITY_STATUS_FROM_SERVER.DRAFT,
  ENTITY_STATUS_FROM_SERVER.UPDATED
];

const listShowBtnSMDelete = [ENTITY_STATUS_FROM_SERVER.DRAFT, ENTITY_STATUS_FROM_SERVER.INACTIVATE];

const listShowBtnDeactivate = [ENTITY_STATUS_FROM_SERVER.ACTIVATE, ENTITY_STATUS_FROM_SERVER.UPDATED];

const listShowBtnDelete = [ENTITY_STATUS_FROM_SERVER.INACTIVATE, ENTITY_STATUS_FROM_SERVER.DRAFT];

export const checkEntityActionNonApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnActivate.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivate.includes(entityStatus);
    case DELETE:
      return listShowBtnDelete.includes(entityStatus);
    default:
      return true;
  }
};

export const checkSMEntityActionNonApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnSMActivate.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivate.includes(entityStatus);
    case DELETE:
      return listShowBtnSMDelete.includes(entityStatus);
    default:
      return true;
  }
};

export const getAllSubGenres = (genres: Genre[]) =>
  genres
    .map(({ children }) => children)
    .reduce((currentValue, subGenres) => {
      if (subGenres) {
        return [...currentValue, ...subGenres];
      }
      return currentValue;
    }, []);

export const getAllMicroGenres = (subGenres: SubGenre[]) =>
  subGenres
    .map(({ children }) => children)
    .reduce((currentValue, microGenre) => {
      if (microGenre) {
        return [...currentValue, ...microGenre];
      }
      return currentValue;
    }, []);

const listShowBtnActivateWithApproval = [...listShowBtnActivate, ENTITY_STATUS_FROM_SERVER.REJECT];
const listShowBtnDeactivateWithApproval = [...listShowBtnDeactivate];
const listShowBtnDeleteWithApproval = [...listShowBtnDelete];
const listShowBtnApproveWithApproval = [ENTITY_STATUS_FROM_SERVER.APPROVAL_PENDING];
const listShowBtnRejectWithApproval = [ENTITY_STATUS_FROM_SERVER.APPROVAL_PENDING];

export const checkEntityActionApproval = (action: string, entityStatus: string) => {
  switch (action) {
    case ACTIVATE:
      return listShowBtnActivateWithApproval.includes(entityStatus);
    case DEACTIVATE:
      return listShowBtnDeactivateWithApproval.includes(entityStatus);
    case DELETE:
      return listShowBtnDeleteWithApproval.includes(entityStatus);
    case APPROVE:
      return listShowBtnApproveWithApproval.includes(entityStatus);
    case REJECT:
      return listShowBtnRejectWithApproval.includes(entityStatus);
    default:
      return true;
  }
};

export function createMarkup() {
  return {
    __html: `<div class="zalo-chat-widget" data-oaid="927003371386666481" data-welcome-message="Rất vui khi được hỗ trợ bạn!" data-autopopup="0" data-width="450" data-height="450"></div>
    `
  };
}
