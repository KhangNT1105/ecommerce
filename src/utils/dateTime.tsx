import {
  format,
  parse,
  getTime,
  subDays,
  subMonths,
  startOfWeek,
  endOfWeek,
  startOfMonth,
  endOfMonth,
  startOfToday,
  endOfToday,
  startOfDay,
  endOfDay
} from 'date-fns';

import { DATE_FORMAT, DATE_TIME_FULL_FORMAT } from 'constants/index';

export type DateRangeType = { from: string | null; to: string | null };

export const formatDate = (timestamp: string) => {
  return format(parse(timestamp, 'T', new Date()), DATE_FORMAT);
};

export type DateRangeProps = {
  dateFrom: string | null;
  dateTo: string | null;
};

/**
 * get timestamp from a string
 * @param: value string
 */
export const getTimestamp = (value: string): number => {
  return getTime(parse(value, DATE_FORMAT, new Date()));
};

export const getUTCDateFn = (date: Date) => {
  return new Date(
    date.getUTCFullYear(),
    date.getUTCMonth(),
    date.getUTCDate(),
    date.getUTCHours(),
    date.getUTCMinutes(),
    date.getUTCSeconds()
  );
};

/**
 * Description: combine day from DatePicker and local time then convert to UTC Date
 * @param date: from date picker
 * @return: UTC Date
 */
export const convertDatePickerToUTC = (date: Date) => {
  const today = new Date();
  const combineDateTime = new Date(
    date.getFullYear(),
    date.getMonth(),
    date.getDate(),
    today.getHours(),
    today.getMinutes(),
    today.getSeconds()
  );

  return getUTCDateFn(combineDateTime);
};

/**
 * Description: convert HH:MM:SS string to seconds
 * @param dateString
 * @param second
 */
export const convertTimeStringToMillisecond = (dateString: String) => {
  return (dateString as any).split(':').reduce((acc: any, time: any) => 60 * acc + +time) * 1000;
};

export const convertDateFromToUTC = (date: Date) => {
  const combineDateTime = new Date(date.getFullYear(), date.getMonth(), date.getDate());

  return getUTCDateFn(startOfDay(combineDateTime));
};

export const convertDateToToUTC = (date: Date) => {
  const combineDateTime = new Date(date.getFullYear(), date.getMonth(), date.getDate());

  return getUTCDateFn(endOfDay(combineDateTime));
};

export const getToday = (): DateRangeProps => {
  const dateFrom = getUTCDateFn(startOfToday());
  const dateTo = getUTCDateFn(endOfToday());

  return {
    dateFrom: format(dateFrom, DATE_TIME_FULL_FORMAT),
    dateTo: format(dateTo, DATE_TIME_FULL_FORMAT)
  };
};

export const getLast7days = (): DateRangeProps => {
  const dateTo = format(getUTCDateFn(endOfToday()), DATE_TIME_FULL_FORMAT);
  const dateFrom = format(getUTCDateFn(subDays(startOfToday(), 7)), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const getLast30days = (): DateRangeProps => {
  const dateTo = format(getUTCDateFn(endOfToday()), DATE_TIME_FULL_FORMAT);
  const dateFrom = format(getUTCDateFn(subDays(startOfToday(), 30)), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const getLastWeek = (): DateRangeProps => {
  const dateFrom = format(getUTCDateFn(startOfWeek(subDays(startOfToday(), 7))), DATE_TIME_FULL_FORMAT);
  const dateTo = format(getUTCDateFn(endOfWeek(subDays(endOfToday(), 7))), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const getLastMonth = (): DateRangeProps => {
  const dateFrom = format(getUTCDateFn(startOfMonth(subMonths(startOfToday(), 1))), DATE_TIME_FULL_FORMAT);
  const dateTo = format(getUTCDateFn(endOfMonth(subMonths(endOfToday(), 1))), DATE_TIME_FULL_FORMAT);

  return { dateFrom, dateTo };
};

export const convertMillisecondToHHMMSS = (milliseconds: number) => {
  if (!milliseconds) return '';
  const date = new Date(milliseconds);
  const hh = date.getUTCHours();
  const mm = date.getUTCMinutes();
  const ss = date.getSeconds();
  const hours = hh < 10 ? `0${hh}` : `${hh}`;
  const minutes = mm < 10 ? `0${mm}` : `${mm}`;
  const seconds = ss < 10 ? `0${ss}` : `${ss}`;

  return `${hours}:${minutes}:${seconds}`;
};
