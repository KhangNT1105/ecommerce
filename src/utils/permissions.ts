import { matchPath } from 'react-router';
import { FieldType, PermissionFieldsType, FormPermissionsType } from '../components/roleFormField/RoleFormFields.d';
import { BUTTONS_VISIBILITY, PERMISSION_PREFIX, PAGE_ACCESSIBILITY } from '../constants/enum';
/**
 * Description: check user permissions have spec permissions
 * @param permissions: array of user permissions
 * @param modes: array of PERMISSION_PREFIX
 * Results: boolean
 */
export const hasModes = (permissions: string[], modes: string[]) => {
  const foundMode = permissions.filter((item) => {
    const includes = modes.filter((mode) => {
      return item === mode;
    });
    return includes.length > 0;
  });
  return foundMode.length === modes.length;
};

/**
 * Description: check user permissions only have 1 permission
 * @param permissions: array of user permissions
 * @param mode: PERMISSION_PREFIX
 * Results: boolean
 */
export const hasOnlyMode = (permissions: string[], mode: string) => {
  const viewMode = permissions.filter((item) => {
    return item === mode;
  });
  return permissions.length === 1 && viewMode.length === 1;
};

/**
 * Description: generate new structure data:
 * Params:
 *    @param formPermissions: array of permissions and element name (id)
 *    @param userPermissions: array of permissions of logged in users
 *
 * Results:
 * {
 *  btnSave: 'show',
 *  btnCancel: 'NA',
 *  ...
 * }
 */
export const reGeneratePermissionFields = (formPermissions: FormPermissionsType[], userPermissions: string[]) => {
  const permissionFields: PermissionFieldsType = formPermissions
    .filter(({ permissions }) => {
      return userPermissions.includes(permissions);
    })
    .map((formPermissions: FormPermissionsType) => {
      const { permissions, ...restPermissions } = formPermissions;
      return restPermissions;
    });

  const fieldsWithPermissions = permissionFields.reduce(
    (accumulator: FieldType, currentValue: FieldType, index: number) => {
      const fields = {
        ...accumulator
      };
      Object.keys(currentValue).map((operation: string) => {
        if (currentValue[operation] === BUTTONS_VISIBILITY.SHOW) {
          fields[operation] = currentValue[operation];
        }
        return true;
      });
      return {
        ...fields
      };
    },
    {}
  );
  return fieldsWithPermissions;
};
/**
 * Description: get accessibility of page with specification page
 * Params:
 *    @param formPermissions: array of permissions of page
 *    @param userPermissions: array of permissions of logged in users
 *    @param strPermission: permission
 *
 * Results:
 * pageAccessibility: boolean
 */
export const checkPageAccessibility = (
  formPermissions: FormPermissionsType[],
  userPermissions: string[],
  strPermission: string
) => {
  const pagePermissions = reGeneratePermissionFields(formPermissions, userPermissions);
  return pagePermissions[strPermission] === PAGE_ACCESSIBILITY.SHOW;
};

/**
 * Description: get accessibility of page with specification entity
 * Params:
 *    @param path: current path of site
 *    @param entityName: entity
 *    @param formPermissionsCreate: array of permissions of create page
 *    @param formPermissionsEdit: array of permissions of edit page
 *    @param formPermissionsList: array of permissions of list page
 *    @param pathCreate: route string of create page
 *    @param pathEdit: route string of edit page
 *    @param userPermissions: array of permissions of logged in users
 *
 * Results:
 * pageAccessibility: boolean
 */
export const getPageAccessibility = (
  path: string,
  entityName: string,
  formPermissionsCreate: FormPermissionsType[],
  formPermissionsEdit: FormPermissionsType[],
  formPermissionsList: FormPermissionsType[],
  pathCreate: string,
  pathEdit: string,
  userPermissions: string[]
) => {
  let pageAccessibility = true;
  if (matchPath(path, { path: pathCreate, exact: true })) {
    pageAccessibility = checkPageAccessibility(
      formPermissionsCreate,
      userPermissions,
      `${entityName}_${PERMISSION_PREFIX.CREATE.toLowerCase()}`
    );
  } else if (matchPath(path, { path: pathEdit, exact: true })) {
    pageAccessibility = checkPageAccessibility(
      formPermissionsEdit,
      userPermissions,
      `${entityName}_${PERMISSION_PREFIX.EDIT.toLowerCase()}`
    );
  } else {
    pageAccessibility = checkPageAccessibility(formPermissionsList, userPermissions, `${entityName}`);
  }
  return pageAccessibility;
};

/**
 * Description: check page is accessible by entity
 * Params:
 *    @param page: page name
 *    @param path: current path of site
 *    @param userPermissions: array of permissions of logged in users
 *
 * Results:
 * pageAccessibility: boolean
 */
export const checkPagePermissisons = (page: string, path: string, userPermissions: string[]) => {
  // let pagePermissions: { [permission: string]: string };
  const pageAccessibility = true;

  return pageAccessibility;
};

/**
 * Description: get form permissions of list page by entity
 * Params:
 *    @param entity: entity name
 *
 * Results:
 * formPermissions: FormPermissionsType[]
 */
export const getFormPermissionsPageList = (entity: string) => {};
