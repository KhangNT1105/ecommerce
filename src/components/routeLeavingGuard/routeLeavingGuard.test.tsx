import * as React from 'react';
import { shallow } from 'enzyme';

import RouteLeavingGuard, { handleBeforeUnload } from './routeLeavingGuard';
import { testSnapshots } from '../../utils/test';
import { PAGE_ENTITY } from 'constants/enum';

jest
  .mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
      push: jest.fn()
    }),
    Prompt: () => <div />
  }))
  .mock('components/modal/WarningMessage/WarningMessage', () => 'WarningMessage')
  .mock('stores/AuthenticationStore/authentication', () => () => [
    {
      user: {
        permissions: ['CREATE_LICENSOR', 'VIEW_LICENSOR', 'EDIT_LICENSOR']
      }
    }
  ]);

describe('<RouteLeavingGuard/>', () => {
  const setState = jest.fn();
  const useStateMock: any = (initState: any) => [initState, setState];
  jest.spyOn(React, 'useState').mockImplementation(useStateMock);
  const wrapper = () => <RouteLeavingGuard when={true} />;

  testSnapshots(wrapper, [
    {
      description: 'render RouteLeavingGuard snapshot',
      props: {
        when: true
      }
    }
  ]);
});

describe('Handlers', () => {
  let wrapper: any;
  let useEffect: any;
  const setState = jest.fn();

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f: any) => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    mockUseEffect();
    mockUseEffect();
    const useStateMock: any = (initState: any) => [initState, setState];
    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
    wrapper = shallow(<RouteLeavingGuard when={true} mode={'/'} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('handleCancel to be called', () => {
    const WarningMessage: any = wrapper.find('WarningMessage').props();
    WarningMessage.onCancelHandler();
    expect(setState).toHaveBeenCalledWith(false);
  });
  test('handleOk to be called', () => {
    const WarningMessage: any = wrapper.find('WarningMessage').props();
    WarningMessage.onOkHandler();
    expect(setState).toHaveBeenCalledWith(false);
    expect(setState).toHaveBeenCalledWith(true);
  });
  test('handleBlockedNavigation to be called with showing popup', () => {
    const event: any = {
      preventDefault: jest.fn(),
      returnValue: ''
    };
    handleBeforeUnload(event);

    const Prompt: any = wrapper.find('Prompt').props();
    Prompt.message('/login');
  });
});
