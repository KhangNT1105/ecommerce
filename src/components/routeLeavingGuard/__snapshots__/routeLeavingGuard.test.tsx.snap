// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RouteLeavingGuard/> snapshots render RouteLeavingGuard snapshot 1`] = `
Array [
  <div />,
  <WarningMessage
    isOpen={false}
    message="LOST_DATA"
    onCancelHandler={[Function]}
    onOkHandler={[Function]}
    textCancel="CANCEL"
    textOk="OK"
    title="WARNING"
  />,
]
`;
